zipline/utils/cache.py

```diff
@@ -86,6 +88,33 @@ def _unsafe_get_value(self):
         return self._value
 
 
+def basic_cleanup(cache, key):
+    del cache[key]
+
+
+def dataframe_gc_cleanup(cache, key):
+    # Try to deterministically garbage collect the previous result by
+    # removing any references to it. There are at least three sources
+    # of references:
+
+    # 1. self._pipeline_cache holds a reference.
+    # 2. The dataframe itself holds a reference via cached .iloc/.loc
+    #    accessors.
+    # 3. The traceback held in sys.exc_info includes stack frames in
+    #    which self._pipeline_cache is a local variable.
+
+    # We remove the above sources of references in reverse order:
+
+    # 3. Clear the traceback.  This is no-op in Python 3.
+    exc_clear()
+
+    # 2. Clear the .loc/.iloc caches.
+    clear_dataframe_indexer_caches(cache[key]._unsafe_get_value())
+
+    # 1. Clear the reference to the `CachedObject`
+    del cache[key]
```

Comment:
I don't think the cleanup function should have to be responsible for clearing the cache key.  That should happen unconditionally no matter what.

I think the right signature for the cleanup function is to just take the value wrapped in the `CachedObject`.